lib-src/rpc2/rp2gen/crout.c

```diff
@@ -185,13 +188,220 @@ static rp2_bool legal_struct_fields[] = {
 	/* RPC2_Byte */		RP2_TRUE,
 	/* RPC2_String */		RP2_TRUE,	/* Untested */
 	/* RPC2_CountedBS */	RP2_TRUE, /* Untested */
-	/* RPC2_BoundedBS */	RP2_TRUE, 
+	/* RPC2_BoundedBS */	RP2_TRUE,
 	/* RPC2_BulkDescriptor */	RP2_FALSE,  /* Untested */
 	/* RPC2_EncryptionKey */	RP2_TRUE,   /* Untested */
 	/* RPC2_Struct */		RP2_TRUE,
 	/* RPC2_Enum */		RP2_TRUE,
 	/* RPC2_Double */		RP2_TRUE
 };
+void print_struct_func(RPC2_TYPE *t, FILE *where, FILE *hfile, char *name) {
+    if (t->tag != RPC2_STRUCT_TAG)
+        return;
+    fprintf(where, "int pack_struct_%s (BUFFER *buf, %s *ptr, RPC2_Integer who) {\n", name, name);
+    fprintf(hfile, "int pack_struct_%s (BUFFER *buf, %s *ptr, RPC2_Integer who);\n", name, name);
+
+    VAR **v;
+	for (v=t->fields.struct_fields; *v!=NIL; v++) {
+        if (!legal_struct_fields[(int32_t) (*v)->type->type->tag]) {
+			printf("RP2GEN: illegal type for RPC2_Struct field: %s\n", (*v)->name);
+			exit(1);
+		}
+		print_pack_var("ptr->", *v, where);
+	}
+    fputs("    return 0;\n", where);
```

Comment:
Instead of mixing `fprintf(where, ...);` and `fputs(..., where);`, it may look better to just use fprintf everywhere. It may be slightly slower as fprintf has to scan the string for `%` escapes, but I don't think we are much about speed of rpc2 stub generation.

The only place we couldn't replace fputs with fprintf is when we are writing strings we don't control that may contain '%' escapes.